### PR TITLE
refactor: upstream Azure managed instance cache refactor

### PR DESCRIPTION
**Pull request by comtalyst**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
"Refactor" as a part of fork-upstream (managed-selfhosted) realignment. Should not have any breaking changes.
This codebases realignment will simplify the logistics between the two, cutting a significant portion of maintenance cost.
This PR specifically focus on instance caching.

There will be a separate effort focusing on improve code quality, rather than realigning codebase.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Most changes are merely refactor. Also include small fixes to refresh instance cache in more cases that should be required. Also include support for currently managed-exclusive flags (e.g., `EnableDetailedCSEMessage`).

There are similar PRs out on version release branches as well. Should be near identical.

#### Does this PR introduce a user-facing change?

```release-note
- Azure: fix some cases where instance cache is outdated but not getting refreshes
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/azure/azure_force_delete_scale_set.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	return future, rerr
+}
+
+func shouldForceDelete(skuName string, scaleSet *ScaleSet) bool {
+	return scaleSet.enableForceDelete && !isolatedVMSizes[strings.ToLower(skuName)] && !scaleSet.dedicatedHost
+}
+
+func isOperationNotAllowed(rerr *retry.Error) bool {
+	return rerr != nil && rerr.ServiceErrorCode() == retry.OperationNotAllowed
+}
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_force_delete_scale_set.go:86]: We should move these to azure sdk for go extensions

**Review by Bryce-Soghigian — APPROVED** [review]
/approve

File: cluster-autoscaler/cloudprovider/azure/azure_force_delete_scale_set.go (hunk 1 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_force_delete_scale_set.go:86]: Do you mean we have an equivalent of this in a library?
Also, I don't think small things like this is necessary to be abstracted in the first place.

But, whatever idea we have, let's save it for post-defork quality improvement. It doesn't look worth it to do that now---in this case.

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ func newTestAzureManager(t *testing.T) *AzureManager {
 			VMType:              vmTypeVMSS,
 			MaxDeploymentsCount: 2,
 			Deployment:          "deployment",
-			EnableForceDelete:   true,
+			EnableForceDelete:   false,
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:59]: why false now by default?

Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: Some unit tests were not synced perfectly apparently. Fixed.

**Review by Bryce-Soghigian — APPROVED** [review]

**Pull request merged**